### ISSUE TITLE
Support Input Handlers in Middleware

### ISSUE THREAD

**Issue by joozek78** [Starcounter/Starcounter#3807]
Middleware currently allows filtering HTTP requests (for example getting a page for the first time), but it doesn't allow to interfere with existing page's behavior.
To add custom behavior reacting to input on a page (such as authorization) one has to create base class, like `SecurePage` and add there a bunch of calls to `Property<T>.AddHandler`. I would like to be able to add a Middleware in similar fashion:

```c#
Application.Use((json, property, value) => { do something });
```

Bonus question: can I filter the requests for specific Page? To act before any json is actually sent?

**Comment by alemoi** [comment]
The API was changed recently. @per-samuelsson can you give some insight on the question? If not, I will check it. Thanks.

**Comment by miyconst** [comment]
@joozek78 I understand your question and concern.

Do you think it would be easier to use attributes instead of filters and custom `SecurePage` classes? I mean something like this:

```cs
[Authorize(MyApp.Roles.Admin, MyApp.Roles.Manager)]
public class MyPage : Json {
    [Authorize(MyApp.Roles.Admin)]
    protected void Handle(Input.DeleteClick Action) {
    }
}
```

This is how it works in ASP.NET and I liked it when used.

**Comment by per-samuelsson** [comment]
> @per-samuelsson can you give some insight on the question?

Sorry, didn't see this until now. Sure I can, I just have to review the code I wrote to remember. :wink: 

I guess the question I can provide an answer for is this:

> Bonus question: can I filter the requests for specific Page? To act before any json is actually sent?

**Comment by per-samuelsson** [comment]
> Bonus question: can I filter the requests for specific Page? To act before any json is actually sent?

@joozek78, something like this?

**Foo.json**
```json
{
  "Value": "Foo"
}
```

**Program.cs**
```cs
class Program
{
    static void Main()
    {
        var app = Application.Current;
        app.Use((req, resp) =>
        {
            var page = resp.Resource as Foo;
            if (page != null)
            {
                page.Value = "Fubar";
            }

            return resp;
        });

        Handle.GET("/foo", () =>
        {
            return new Foo() { Value = "Bar" };
        });
    }
}
```

**http://localhost:8080/foo**
```
{"Value":"Fubar"}
```

**Comment by joozek78** [comment]
@per-samuelsson thanks for answer. Unfortunately, my question wasn't precise :) Let me rephrase it:

> can I filter the requests for specific Page? To act before any ~~json is actually sent~~ Page code is executed?
 
@miyconst I do think it would be a cleaner solution. Base class is a work-around, because middleware I described is not yet implemented

**Comment by per-samuelsson** [comment]
> can I filter the requests for specific Page? To act before any ~~json is actually sent~~ Page code is executed?

Then I'm not sure exactly what you mean I'm afraid. Do you have any pseudo code you could provide that illustrate what you ultimately would want?

The closest I can think of is this really trivial pattern, using a request filter. But I'm not at all very knowledgeable in this area. So it's probably not what you are after.

```cs
app.Use((req) =>
{
    if (req.Uri == "/foo")
    {
        return new Foo();
    }

    // Effectively invoking handler
    return null;
});

Handle.GET("/foo", () => {
    // I will never get called :(
    return new Foo() { Value = "Bar" };
});
```

**Comment by joozek78** [comment]
I feel that I unwittingly spawned an off-topic discussion here. I'll extract it to a separate issue.
Back to the main topic:
**Can we expect support for Input handlers in middleware, or are there any problems that prevent it from happening?**

**Comment by joozek78** [comment]
Okay, I was able to work around and provide this functionality with pure app code. It will be included (probably?) in PageLoader, which will be developed initially in sweoffshore project.
